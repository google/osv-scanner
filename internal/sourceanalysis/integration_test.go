package sourceanalysis

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/google/osv-scanner/v2/internal/testutility"
	"github.com/ossf/osv-schema/bindings/go/osvschema"
)

var fixturesDir = "integration/fixtures-go"

func Test_runGovulncheck(t *testing.T) {
	t.Parallel()
	entries, err := os.ReadDir(fixturesDir)
	if err != nil {
		t.Errorf("failed to read fixtures dir: %v", err)
	}

	vulns := []osvschema.Vulnerability{}
	for _, de := range entries {
		if !de.Type().IsRegular() {
			continue
		}

		if !strings.HasSuffix(de.Name(), ".json") {
			continue
		}

		fn := filepath.Join(fixturesDir, de.Name())
		file, err := os.Open(fn)
		if err != nil {
			t.Errorf("failed to open fixture vuln files: %v", err)
		}

		newVuln := osvschema.Vulnerability{}
		err = json.NewDecoder(file).Decode(&newVuln)
		if err != nil {
			t.Errorf("failed to decode fixture vuln file (%q): %v", fn, err)
		}
		vulns = append(vulns, newVuln)
	}

	res, err := runGovulncheck(filepath.Join(fixturesDir, "test-project"), vulns, "1.19")
	if err != nil {
		t.Errorf("failed to run RunGoVulnCheck: %v", err)
	}

	res["GO-2023-1558"][2].Trace[0].Position.Filename = "<Any value>"
	res["GO-2023-1558"][2].Trace[1].Position.Filename = "<Any value>"
	res["GO-2023-1558"][2].Trace[0].Position.Offset = -1
	res["GO-2023-1558"][2].Trace[1].Position.Offset = -1

	for _, traceItem := range res["GO-2023-2382"][2].Trace {
		traceItem.Position.Filename = "<Any value>"
		traceItem.Position.Offset = -1
		traceItem.Position.Line = -1 // This number differs between go versions

		if traceItem.Function == "ListenAndServe" && traceItem.Receiver == "*Server" {
			traceItem.Position.Column = -1 // This number differs between go versions
		}
	}

	testutility.NewSnapshot().MatchJSON(t, res)
}
