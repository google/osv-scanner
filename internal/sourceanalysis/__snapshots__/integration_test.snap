[Test_runGovulncheck - 1]
{
  "GO-2021-0053": [
    {
      "osv": "GO-2021-0053",
      "fixed_version": "v1.3.2",
      "trace": [
        {
          "module": "github.com/gogo/protobuf",
          "version": "v1.3.1"
        }
      ]
    },
    {
      "osv": "GO-2021-0053",
      "fixed_version": "v1.3.2",
      "trace": [
        {
          "module": "github.com/gogo/protobuf",
          "version": "v1.3.1",
          "package": "github.com/gogo/protobuf/plugin/unmarshal"
        }
      ]
    }
  ],
  "GO-2023-1558": [
    {
      "osv": "GO-2023-1558",
      "fixed_version": "v1.1.0",
      "trace": [
        {
          "module": "github.com/ipfs/go-bitfield",
          "version": "v1.0.0"
        }
      ]
    },
    {
      "osv": "GO-2023-1558",
      "fixed_version": "v1.1.0",
      "trace": [
        {
          "module": "github.com/ipfs/go-bitfield",
          "version": "v1.0.0",
          "package": "github.com/ipfs/go-bitfield"
        }
      ]
    },
    {
      "osv": "GO-2023-1558",
      "fixed_version": "v1.1.0",
      "trace": [
        {
          "module": "github.com/ipfs/go-bitfield",
          "version": "v1.0.0",
          "package": "github.com/ipfs/go-bitfield",
          "function": "NewBitfield",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": 12,
            "column": 6
          }
        },
        {
          "module": "github.com/ossf-tests/osv-e2e",
          "package": "github.com/ossf-tests/osv-e2e",
          "function": "main",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": 16,
            "column": 22
          }
        }
      ]
    }
  ],
  "GO-2023-2382": [
    {
      "osv": "GO-2023-2382",
      "fixed_version": "v1.20.12",
      "trace": [
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http/internal"
        }
      ]
    },
    {
      "osv": "GO-2023-2382",
      "fixed_version": "v1.20.12",
      "trace": [
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http/internal"
        }
      ]
    },
    {
      "osv": "GO-2023-2382",
      "fixed_version": "v1.20.12",
      "trace": [
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http/internal",
          "function": "Read",
          "receiver": "*chunkedReader",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 26
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "readLocked",
          "receiver": "*body",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 21
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "Read",
          "receiver": "bodyLocked",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 24
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "io",
          "function": "copyBuffer",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 21
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "io",
          "function": "Copy",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 19
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "Close",
          "receiver": "*body",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 19
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "finishRequest",
          "receiver": "*response",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 17
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "serve",
          "receiver": "*conn",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 18
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "Serve",
          "receiver": "*Server",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 3
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "ListenAndServe",
          "receiver": "*Server",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 18
          }
        },
        {
          "module": "stdlib",
          "version": "v1.19.0",
          "package": "net/http",
          "function": "ListenAndServe",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 30
          }
        },
        {
          "module": "github.com/ossf-tests/osv-e2e",
          "package": "github.com/ossf-tests/osv-e2e",
          "function": "main",
          "position": {
            "filename": "\u003cAny value\u003e",
            "offset": -1,
            "line": -1,
            "column": 28
          }
        }
      ]
    }
  ]
}
---
